rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,258 @@
+[removing an endpoint](#removing-an-endpoint) guidelines for removing fields.
+
+Bad
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // ContainsDuplicates bool <--- previously existing field made redundant by DuplicateCount
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+         // ContainsDuplicates must be populated to support older clients.
+	     DuplicateCount: 1 
+	}, nil
+}
+```
+
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // TODO(avatus): DELETE IN 18. Use `DuplicateCount` instead.
+    ContainsDuplicates bool
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
+	     ContainsDuplicates: true
+	     DuplicateCount: 1 
+	}, nil
+}
+``` 
+If a response shape must change entirely, prefer creating a new endpoint
+
+### Updating request body fields
+A Request body can be updated with a new field as long as the omission of the
+new field would not cause a breaking change (i.e., an old client sending the old
+request without the new field). If the handler _requires_ this new field in
+order to function, then a new endpoint must be created with a version increment.
+
+The existing request adding in the new `MyNewThing` field
+```go
+type MyRequest struct {
+    Limit int
+    Query string
+    MyNewThing string // <-- new field
+}
+```
+
+if the handler has code along the lines of this
+```go
+if req.MyNewThing == "" {
+     return nil, err("BAD REQUEST")
+}
+```
+then that is a *BREAKING CHANGE* and should thus have a new versioned handler.
+
+
+
+### Adding new features to the web client
+If a feature requires a new endpoint, it must properly handle the case of a 404.
+This means making the "expected" 404 error clear to the user that an endpoint is
+not found due to a version mismatch (and preferably which minimum version is
+needed) rather than an ambiguous "Not Found" error. This works mostly for a
+"resources" endpoint where it is expected to return an empty list if something
+isn't found (i.e., "no access requests"), but a new endpoint that returns a
+single resource _should not_ assume that a 404 means anything other than a
+resource is not found. In this particular case, there isn't much graceful work
+we can do here. Graceful handling of this error state should be best effort.
+
+### New required fields from an API response in the web client
+If the updated feature _cannot_ function without receiving new fields from the
+API (for example, receiving a response from a proxy version N-1), refer to the
+API guidelines about creating a new endpoint. If the feature itself is degraded
```

Comment:
Nit: the guideline you are referring to is about creating an endpoint that didn't exist before, not about the version increase.